### PR TITLE
Minesweeper class

### PR DESCRIPTION
**Pull request by dariannax3**

### CODE REVIEW

File: src/Minesweeper.cpp [hunk 1]
```diff
@@ -1,3 +1,107 @@
...
+    viewer.drawBoard(gameboard_);
+  }
+
+  if (getGameStatus() == GameStatus::win) {
+    viewer.printWin();
+  }
+  if (getGameStatus() == GameStatus::gameOver) {
+    viewer.printLose();
+  }
+}
```
Comment (Rosein) [inline src/Minesweeper.cpp:107]: Missing newline at the end of file.

File: src/Minesweeper.cpp [hunk 2]
```diff
@@ -1,3 +1,107 @@
...
+void Minesweeper::play() {
+  std::cout << "\nWelcome in Minesweeper!\n To choose a field type: action "
+               "position_x position y\n"
+            << std::endl;
+
+  StdViewer viewer;
+  generateBoard();
+  viewer.drawBoard(gameboard_);
+
+  std::cout << "> ";
```
Comment (Rosein) [inline src/Minesweeper.cpp:94]: Magic constant string. It should be extract to some constant variable name something like kPrompt

File: src/Viewer.cpp [hunk 3]
```diff
@@ -3,19 +3,34 @@
 void StdViewer::drawBoard(const Gameboard& gameboard) {
   for (int i = 0; i < gameboard.getWidth(); i++) {
     for (int j = 0; j < gameboard.getHeight(); j++) {
-      if (gameboard.getFieldAt(i, j).flagability == Flagability::marked) {
-        std::cout << " F";
-      } else if (gameboard.getFieldAt(i, j).visibility == Visibility::covered) {
-        std::cout << " X";
+      if (gameboard.getFieldAt(i, j).visibility == Visibility::covered) {
+        if (gameboard.getFieldAt(i, j).flagability == Flagability::marked) {
+          std::cout << " F";
```
Comment (Rosein) [inline src/Viewer.cpp:8]: Every literal magic symbol should be extracted to some const variable with meaningful name

File: src/Viewer.cpp [hunk 4]
```diff
@@ -3,19 +3,34 @@
...
   std::this_thread::sleep_for(std::chrono::microseconds(10000));
+}
+
+void StdViewer::printWin() {
+  std::cout << "Congratulation! You won!" << std::endl;
+}
+
+void StdViewer::printLose() {
+  std::cout << "You lose! Try again" << std::endl;
 }
```
Comment (Rosein) [inline src/Viewer.cpp:36]: Missing new line at the end of file.

File: tests/GameboardTest.cpp [hunk 5]
```diff
@@ -131,4 +131,36 @@ TEST_F(
...
+  gameboard.uncoverOneField(1,2);
+  EXPECT_TRUE(gameboard.getFieldAt(1, 3).visibility == Visibility::covered);
+}
+
+TEST_F(GameboardTestFixture, givenBoardWithOneBombAndOneFlagWhenClickEmptyFieldsThanUncoverAllFieldsWithoutFlagged) {
+  gameboard.setBombAt(0,0);
+  gameboard.flagField(6,6);
+  gameboard.uncoverOneField(1,2);
+  EXPECT_TRUE(gameboard.getFieldAt(6, 6).visibility == Visibility::covered);
 }
```
Comment (Rosein) [inline tests/GameboardTest.cpp:166]: Missing new line at the end of file.

File: tests/MinesweeperTest.cpp [hunk 6]
```diff
@@ -22,4 +27,120 @@ TEST(MinesweeperTest, MakeMovesTest) {
       .WillOnce(Return(std::make_tuple<char, int, int>('w', 1, 1)));
   EXPECT_EQ(player.makeMove(), (std::make_tuple<char, int, int>('w', 1, 1)));
   auto something = player.makeMove();
+}
+
+TEST(MinesweeperTest, givenBoardWhenGenerateBoardThanShouldGenerateTenBombs) {
+  Minesweeper minesweeper(std::make_shared<MockPlayerInput>(),
+                          std::make_shared<StdRandomRangeGenerator>(0, 8));
```
Comment (Rosein) [inline tests/MinesweeperTest.cpp:34]: Magic numbers.

**Review by Rosein — CHANGES_REQUESTED** [review]
Good job, please refactor to the remarks and keep good work up

File: inc/Minesweeper.hpp [hunk 7]
```diff
@@ -1,17 +1,38 @@
...
  private:
-  PlayerPtr player;
+  Bombs generateBombs();
+  void setBombsAtBoard(Bombs& generatedBombs);
+  PlayerInputPtr player_;
+  RandomRangeGeneratorPtr generator_;
+  Gameboard gameboard_;
+  int bombsAmount_{10};
+  GameStatus gameStatus_ {GameStatus::ongoing};
 };
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply inc/Minesweeper.hpp:38]: done

File: inc/PlayerInput.hpp [hunk 8]
```diff
@@ -3,6 +3,10 @@
 #include <memory>
 #include <tuple>
 
+constexpr char CHOOSE = 'c';
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply inc/PlayerInput.hpp:6]: done

File: inc/Viewer.hpp [hunk 9]
```diff
@@ -8,10 +8,14 @@
...
+  virtual void printWin() = 0;
+  virtual void printLose() = 0;
   virtual ~ViewerI() {}
 };
 
 class StdViewer : public ViewerI {
  public:
   void drawBoard(const Gameboard& gameboard) override;
+  void printWin();
+  void printLose();
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply inc/Viewer.hpp:20]: done

File: src/Gameboard.cpp [hunk 10]
```diff
@@ -102,4 +102,27 @@ void Gameboard::setBombAt(const int x, const int y) {
 void Gameboard::flagField(const int x, const int y) {
   Field& field = getFieldAt(x, y);
   field.flagability = Flagability::marked;
-};
\ No newline at end of file
+};
+
+void Gameboard::unflagField(const int x, const int y) {
+  Field& field = getFieldAt(x, y);
+  if(field.flagability != Flagability::marked)
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply src/Gameboard.cpp:109]: This case is handled by the Viewer side.
If statement ensures that the field can only be flagged if it is covered.

File: src/Gameboard.cpp [hunk 11]
```diff
@@ -102,4 +102,27 @@ void Gameboard::setBombAt(const int x, const int y) {
...
+   for (auto& row : board_) {
+    for (auto& field : row) {
+      if(field.visibility == Visibility::covered)
+      {
+        counter++;
+      }
+    }
+  }
+  return counter;
+}
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply src/Gameboard.cpp:128]: done

File: src/Minesweeper.cpp (hunk 2 above)
Comment (dariannax3) [reply src/Minesweeper.cpp:94]: done

File: src/Viewer.cpp (hunk 3 above)
Comment (dariannax3) [reply src/Viewer.cpp:8]: done

File: src/Viewer.cpp (hunk 4 above)
Comment (dariannax3) [reply src/Viewer.cpp:36]: done

File: tests/GameboardTest.cpp (hunk 5 above)
Comment (dariannax3) [reply tests/GameboardTest.cpp:166]: done

File: inc/Gameboard.hpp [hunk 12]
```diff
@@ -15,7 +13,7 @@ struct Field {
 };
 
 using Board = std::array<std::array<Field, 8>, 8>;
```
Comment (Rosein) [inline inc/Gameboard.hpp:15]: Two magic numbers

**Review by Rosein — APPROVED** [review]

File: tests/MinesweeperTest.cpp (hunk 6 above)
Comment (dariannax3) [reply tests/MinesweeperTest.cpp]: done

File: inc/Gameboard.hpp (hunk 12 above)
Comment (dariannax3) [reply inc/Gameboard.hpp:15]: done

**Review by Rosein — APPROVED** [review]
Good job!

**Pull request merged**